homeassistant/auth.py

```diff
@@ -121,6 +317,77 @@ def async_create_credentials(self, data):
         return {}
 
 
+class LoginFlow(data_entry_flow.FlowHandler):
+    """Handler for the login flow."""
+
+    def __init__(self, auth_provider):
+        """Initialize the login flow."""
+        self._auth_provider = auth_provider
+        # self._auth_modules is mutable, we need a copy
+        self._auth_modules = auth_provider.modules.copy()
+        self._data = dict()
+
+    async def async_step_init(self, user_input=None):
+        """Handle the first step of login flow.
+
+        Return self.async_show_form(step_id='init') if user_input == None.
+        Return await self.async_finish(username) if login init step pass.
+        """
+        raise NotImplementedError
+
+    async def async_finish(self, username):
+        """Handle the pass of login flow."""
+        if self._auth_modules:
+            _, auth_module = self._auth_modules.popitem(False)
```

Comment:
This is weird, instead, once finish called, we should check if user has any 2fa configured (stored on user object?) and then create a local list of 2fa to ask which we can mutate.

Reply:
A user can have multiple 2FA active, it only needs to pass 1 to log in.

Reply:
I see your point, so the flow should like

admin create user -> user login -> user enable/setup one of many 2fa from a list
user select auth_provider -> auth_provider validate -> user select which 2fa he/she want to use -> ask 2fa challenge -> 2fa module validate -> user logged in

Reply:
Yep, and we can do it all within a login flow without needed a frontend change